### PR TITLE
[Feature 13, 50, and 75] Chat back-end and persistence implemented with SQLite 

### LINKED ISSUE #13: [Feature 13] Implement back-end for sending messages
Implement back-end functionality so a user can send a message to another user.

### LINKED ISSUE #50: [Feature 50] Connect one-on-one chat to SQLite database
Connect one-on-one chat to SQLite database

### LINKED ISSUE #154: [Feature 75] Chat persistence

### PR DESCRIPTION
**Pull request by navarhontes**
Chat now connected to back end and persists correctly. 
* Used a variation on the observer pattern even though that's overengineering for a case we're not actually using, because I said I would. MessageModel array lists as the observable; Presenter as the observer class. 
* ChatInteractor implements ChatPresenter, ChatActivityController
* Replaced previous loadMessages method with ChatPresenter logic 

Some bugs fixed in other files/usecases: 
* findConversationId should use AND, not OR 
* If a conversation should be created whenever a match is created, then all matches should be prepopulated with corresponding conversations. 

Remaining related issues: 
* #52 
* #31 
* #32

Closes the following:
* Closes #13 
* Closes #50 
* Closes #154

### CODE REVIEW

**Review by candidcalcifer — APPROVED** [review]

File: app/src/main/java/com/courseproject/tindar/ui/chat/ChatActivity.java [hunk 1]
```diff
@@ -13,11 +13,18 @@
 import android.widget.TextView;
 
 import com.courseproject.tindar.R;
+
+import com.courseproject.tindar.ds.DatabaseHelper;
 import com.courseproject.tindar.entities.MessageModel;
+
 // TODO: remove TindarMessage import when database is properly connected
```
Comment (jenniferjinyoungyang) [inline app/src/main/java/com/courseproject/tindar/ui/chat/ChatActivity.java:20]: is it still TODO?

File: app/src/main/java/com/courseproject/tindar/ui/chat/ChatActivity.java [hunk 2]
```diff
@@ -13,11 +13,18 @@
...
 import com.courseproject.tindar.entities.TindarMessage;
+import com.courseproject.tindar.usecases.chat.ChatActivityController;
+import com.courseproject.tindar.usecases.chat.ChatInteractor;
+import com.courseproject.tindar.usecases.chat.ChatPresenter;
+import com.courseproject.tindar.usecases.chat.ChatRequestModel;
 
-// TODO: consider removing Timdstamp import when database is properly connected
+// TODO: consider removing Timestamp import when database is properly connected
```
Comment (jenniferjinyoungyang) [inline app/src/main/java/com/courseproject/tindar/ui/chat/ChatActivity.java:27]: is it still TODO?

File: app/src/main/java/com/courseproject/tindar/ui/chat/ChatActivity.java [hunk 3]
```diff
@@ -38,10 +45,9 @@
...
     /*
         TODO:
-        - un-hardcode "Username" to be the actual conversation partner's display name
-        - un-hardcode userId to be the current user's userId
+        - display conversation partner's display name and not userID
         - possibly reverse the message list implementation?
-        - userId, otherUserID, conversationPartnerDisplayName, and probably a few other things
+        - userID, otherUserID, conversationPartnerDisplayName, and probably a few other things
             should probably be placed in a facade or something
      */
```
Comment (jenniferjinyoungyang) [inline app/src/main/java/com/courseproject/tindar/ui/chat/ChatActivity.java:52]: Is it still TOD?

File: app/src/main/java/com/courseproject/tindar/ui/chat/ChatActivity.java [hunk 4]
```diff
@@ -63,16 +70,11 @@ public class ChatActivity extends AppCompatActivity {
     /** List of messages that are already loaded and ready to be displayed by RecyclerView. */
     private ArrayList<MessageModel> loadedMessages;
 
+    /** Chat interactor handling user inputs */
+    private ChatInteractor chatInteractor;
```
Comment (jenniferjinyoungyang) [inline app/src/main/java/com/courseproject/tindar/ui/chat/ChatActivity.java:74]: You should use interface. ChatInputBoundary

File: app/src/main/java/com/courseproject/tindar/usecases/chat/ChatInteractor.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
+package com.courseproject.tindar.usecases.chat;
+
+import com.courseproject.tindar.ds.DatabaseHelper;
+import com.courseproject.tindar.entities.MessageModel;
+
+import java.util.ArrayList;
+
+public class ChatInteractor implements ChatPresenter, ChatActivityController{
```
Comment (jenniferjinyoungyang) [inline app/src/main/java/com/courseproject/tindar/usecases/chat/ChatInteractor.java:8]: ChatInteractor implements ChatInputBoundary (interface for that) and ChatPresenter and ChatActivityController should be dependency injected when instantiating ChatInteractor

File: app/src/main/java/com/courseproject/tindar/usecases/chat/ChatInteractor.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+        if(Integer.parseInt(this.userId) < Integer.parseInt(this.conversationPartnerUserId)){
+            this.userIds = new String[]{userId, conversationPartnerUserId};
+        } else {
+            this.userIds = new String[]{conversationPartnerUserId, userId};
+        }
+
+        this.conversationId = this.chatDsHelper.findConversationId(this.userIds[0], this.userIds[1]);
+        this.updateMessageList();
+        this.observers.add(this);
```
Comment (jenniferjinyoungyang) [inline app/src/main/java/com/courseproject/tindar/usecases/chat/ChatInteractor.java:33]: line 25-33 should go into some method

File: app/src/main/java/com/courseproject/tindar/usecases/chat/ChatInteractor.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+public class ChatInteractor implements ChatPresenter, ChatActivityController{
+    private final ChatDsGateway chatDsHelper;
+    private final String userId;
+    private String conversationPartnerUserId;
+    private String[] userIds;
+    private String conversationId;
+    private ArrayList<MessageModel> messageList;
+    private final ArrayList<ChatPresenter> observers;
+
+    public ChatInteractor(ChatDsGateway chatDsHelper, String userId,
```
Comment (jenniferjinyoungyang) [inline app/src/main/java/com/courseproject/tindar/usecases/chat/ChatInteractor.java:17]: Interactor is just for interacting with other dependency injection. userId or other stuff can be handled in the method

File: app/src/main/java/com/courseproject/tindar/usecases/chat/ChatPresenter.java [hunk 8]
```diff
@@ -0,0 +1,28 @@
+package com.courseproject.tindar.usecases.chat;
+
+import com.courseproject.tindar.entities.MessageModel;
+
+import java.util.ArrayList;
+
+/**
+ * Creates and maintains an ArrayList of MessageModels for ChatActivity to display.
+ */
+public interface ChatPresenter {
```
Comment (jenniferjinyoungyang) [inline app/src/main/java/com/courseproject/tindar/usecases/chat/ChatPresenter.java:10]: This is not presenter. This is more ChatInputBoundary - interface for the interactor

**Comment by jenniferjinyoungyang** [comment]
needs unit testing

**Comment by jenniferjinyoungyang** [comment]
For the interest of time, let's just merge this first. And we can deal with the comments and unit testing in a separate PR.

**Review by jenniferjinyoungyang — APPROVED** [review]

**Review by kresimirgotovac — APPROVED** [review]

**Pull request merged**